src/aapns/pool.py

```diff
@@ -0,0 +1,219 @@
+
+            while len(self.conn) > self.size:
+                c = self.conn.pop()
+                c.closing = True
+                self.dying.add(c)
+                self.cert_hook(c)
+
+            for c in list(self.dying):
+                if c.closed:
+                    self.dying.remove(c)
+                    self.cert_hook(c)
+                elif not c.channels:
+                    self.dying.remove(c)
+                    try:
+                        await c.__aexit__(None, None, None)
+                    finally:
+                        self.cert_hook(c)
+                if self.closing or self.closed:
+                    return
+
+            while len(self.conn) < self.size:
+                if not await self.add_one_connection():
+                    break
+                if self.closing or self.closed:
+                    return
+
+            # FIXME wait for a trigger:
+            # * some connection state has changed
+            with suppress(TimeoutError):
+                await wait_for(self._size_event.wait(), timeout=1)
+            self._size_event.clear()
+
+    async def add_one_connection(self):
+        try:
+            c = Connection(self.origin, ssl=self.ssl)
+            await c.__aenter__()
+            self.conn.add(c)
+            self.cert_hook(c)
+            return True
+        except Exception:
+            logger.exception("Failed creating APN connection")
+            self.cert_hook(c)
+
+    async def __aenter__(self):
+        await gather(*(self.add_one_connection() for i in range(self.size)))
+        self.bg = create_task(self.background_resize(), name="bg-resize")
+        self.started = True
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        self.closing = True
+        if not self.outcome:
+            self.outcome = str(exc) or "Exit"
+        try:
+            if self.bg:
+                self.bg.cancel()
+                with suppress(CancelledError):
+                    await self.bg
+
+            await gather(
+                *(c.__aexit__(exc_type, exc, tb) for c in self.conn),
+                *(c.__aexit__(exc_type, exc, tb) for c in self.dying),
+            )
+        finally:
+            self.closed = True
+
+    async def post(self, req: "Request") -> "Response":
+        with self.count_requests():
+            for delay in (10 ** i for i in count(-3, 0.5)):
+                if self.closing:
+                    raise Closed(self.outcome)
+
+                try:
+                    return await self.post_once(req)
+                except Blocked:
+                    pass
+
+                if self.closing:
+                    raise Closed(self.outcome)
+
+                if time() + delay > req.deadline:
+                    raise Timeout()
+
+                try:
+                    self.retrying += 1
+                    await sleep(delay)
+                finally:
+                    self.retrying -= 1
+            else:
```

Comment:
no need for else here. you can just raise after the `for`

Reply:
I have a feeling that this `else:` can never happen... 

Because `for ... in ... count()` will go on forever (well until `return` or `raise`).
This must be some leftover code.